src/main/java/com/google/neighborgood/servlets/EditTaskServlet.java

```diff
@@ -66,12 +47,65 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
       return;
     }
 
-    // Set the details and rewards to the newly input value
-    task.setProperty("detail", taskDetail);
-    task.setProperty("reward", rewardPts);
-    datastore.put(task);
+    UserService userService = UserServiceFactory.getUserService();
+    if (!userService.isUserLoggedIn()) {
+      System.err.println("User must be logged in to edit a task");
+      response.sendError(
+          HttpServletResponse.SC_UNAUTHORIZED,
+          "You must be logged in to perform this action on a task");
+    }
+
+    // Edits tasks that have been claimed by setting the "helper" property to the userId
+    // of the helper and changing the task's status to "IN PROGRESS"
+    if (request.getParameter("action").equals("helpout")) {
+
+      if (!task.getProperty("status").equals("OPEN")) {
+        System.err.println("Task must be open to be claimed by a helper");
+        response.sendError(
+            HttpServletResponse.SC_CONFLICT, "Task has already been claimed by another helper");
+      }
+
+      String userId = userService.getCurrentUser().getUserId();
```

Comment:
Is there a potential race condition here if another user claims the task in between when the task is queried on line 43 and this change committed on line 71?  Does that mean that the first person would actually be the loser and the second person would end up with the task?  (I don't know anything about how datastore works, so it may have some transaction semantics that protect against this problem that I'm unaware of.)

Reply:
Thanks for pointing this out! I hadn’t heard of race conditions before but it does seem like this is an issue in this situation. It appears that in Datastore you can use Transactions to prevent this, so I’ll implement these changes to handle this edge case on the next commit.